Source/OCMock/NSInvocation+OCMAdditions.m

```diff
@@ -41,16 +43,74 @@ + (NSInvocation *)invocationForBlock:(id)block withArguments:(NSArray *)argument
 }
 
 
-- (BOOL)hasCharPointerArgument
+static NSString *const OCMRetainedObjectArgumentsKey = @"OCMRetainedObjectArgumentsKey";
+
+- (void)retainObjectArgumentsExcluding:(id)objectToExclude
 {
-    NSMethodSignature *signature = [self methodSignature];
-    for(NSUInteger i = 0; i < [signature numberOfArguments]; i++)
+    if(objc_getAssociatedObject(self, OCMRetainedObjectArgumentsKey) == nil)
     {
-        const char *argType = OCMTypeWithoutQualifiers([signature getArgumentTypeAtIndex:i]);
-        if(strcmp(argType, "*") == 0)
-            return YES;
+        NSMutableArray *retainedArguments = [[NSMutableArray alloc] init];
+        NSMethodSignature *methodSignature = self.methodSignature;
+        
+        id target = self.target;
+        if((target != nil) && (target != objectToExclude) && !object_isClass(target))
+        {
+            // Bad things will happen if the target is a block since it's not being
+            // copied. There isn't a very good way to tell if an invocation's target
+            // is a block though (the argument type at index 0 is always "@" even if
+            // the target is a Class or block), and in practice it's OK since you
+            // can't mock a block.
+            [retainedArguments addObject:target];
```

Comment:
I think if you have code compiled with non-ARC blocks can be very different encoded types.  Even with ARC I have seen some differences, I think.  Maybe they have become more consistent on newer compilers but I have seen inconsistencies in the past (and OCMock could be running on lots of setups).  I think dispatch_block_t may be encoded very differently, and it's also easily possible to pass a block into a method argument which is typed to accept any object.  The common non-NSObject block superclass, while an implementation detail, is highly unlikely to change, and is a runtime check of the actual instance -- it should be much safer.